pkg/attachutils/simpleattach.go

```diff
@@ -0,0 +1,385 @@
+
+	// Make the connection and hijack it
+	client.Do(req)
+	conn, br := client.Hijack()
+
+	return conn, br, nil
+}
+
+// CloseWriter is an interface that implement structs
+// that close input streams to prevent from writing.
+type CloseWriter interface {
+	CloseWrite() error
+}
+
+// Code copied from Docker and modified.  stdin, stdout, stderr are the client
+// (hijacked connection) and conn, bufReader are the server
+func AttachStreamsToConn(openStdin, stdinOnce, tty bool, stdin io.ReadCloser, stdout io.Writer, stderr io.Writer, conn net.Conn, bufReader *bufio.Reader, keys []byte) error {
+	var (
+		wg     sync.WaitGroup
+		errors = make(chan error, 3)
+	)
+
+	if stdin != nil && openStdin {
+		wg.Add(1)
+	}
+
+	if stdout != nil {
+		wg.Add(1)
+	}
+
+	if stderr != nil {
+		wg.Add(1)
+	}
+
+	// Connect stdin of container to the http conn.
+	go func() {
+		if stdin == nil || !openStdin {
+			return
+		}
+		log.Debugf("attach: stdin: begin")
+
+		var err error
+		if tty {
+			_, err = copyEscapable(conn, stdin, keys)
+		} else {
+			_, err = io.Copy(conn, stdin)
+		}
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Printf("attach: stdin: %s\n", err)
+			errors <- err
+		}
+		if stdinOnce && !tty {
+			conn.Close()
+		} else {
+			// No matter what, when stdin is closed (io.Copy unblock), close stdout and stderr
+			if stdout != nil {
+				conn.Close()
+			}
+			if stderr != nil {
+				conn.Close()
+			}
+		}
+		log.Debugf("attach: stdin: end")
+		wg.Done()
+	}()
+
+	attachStream := func(name string, stream io.Writer) {
+		if stream == nil {
+			return
+		}
+
+		log.Debugf("attach: %s: begin", name)
+		_, err := io.Copy(stream, bufReader)
+		if err == io.ErrClosedPipe {
+			err = nil
+		}
+		if err != nil {
+			log.Errorf("attach: %s: %v", name, err)
+			errors <- err
+		}
+		// Make sure stdin gets closed
+		if stdin != nil {
+			stdin.Close()
+		}
+		conn.Close()
+		log.Debugf("attach: %s: end", name)
+		wg.Done()
+	}
+
+	go attachStream("stdout", stdout)
+	go attachStream("stderr", stderr)
+
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	<-done
+	close(errors)
+	for err := range errors {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Code copied from Docker and modified.  client streams belong to the hijacked
+// connection and server streams belong to the tether ssh server
+func AttachStreams(tty bool, clientStdin io.ReadCloser, clientStdout, clientStderr io.Writer, svrStdin io.WriteCloser, svrStdout io.Reader, svrStderr io.Reader, keys []byte) error {
+	var (
+		wg     sync.WaitGroup
+		errors = make(chan error, 3)
+	)
+
+	if clientStdin != nil {
+		wg.Add(1)
+	}
+
+	if clientStdout != nil {
+		wg.Add(1)
+	}
+
+	if clientStderr != nil {
+		wg.Add(1)
+	}
+
+	// Connect stdin of container to the http conn.
+	go func() {
+		if clientStdin == nil {
+			return
+		}
+		log.Debugf("attach: stdin: begin")
+
+		var err error
+		if tty {
```

Comment:
The `tty` being allocated on the container happens during `create` in the exec layer.  In effect, we never need to worry about this in the portlayer attach code. 

This means copyEscapable can be removed the PL.